src/rails.js

```diff
@@ -400,73 +439,76 @@
       }
       return false;
-    });
+    },
 
-    $document.delegate(rails.inputChangeSelector, 'change.rails', function(e) {
-      var link = $(this);
+    clickInput: function(e){
+      var button = $(e.currentTarget);
+
+      if (!rails.allowAction(button)) return rails.stopEverything(e);
+
+      // register the pressed submit button
+      var name = button.attr('name'),
+        data = name ? {name:name, value:button.val()} : null;
+
+      button.closest('form').data('ujs:submit-button', data);
+    },
+
+    change: function(e){
+      var link = $(e.currentTarget);
       if (!rails.allowAction(link)) return rails.stopEverything(e);
 
       rails.handleRemote(link);
       return false;
-    });
-
-    $document.delegate(rails.formSubmitSelector, 'submit.rails', function(e) {
-      var form = $(this),
-        remote = form.data('remote') !== undefined,
-        blankRequiredInputs,
-        nonBlankFileInputs;
-
-      if (!rails.allowAction(form)) return rails.stopEverything(e);
+    }
+  }
 
-      // skip other logic when required values are missing or file upload is present
-      if (form.attr('novalidate') == undefined) {
-        blankRequiredInputs = rails.blankInputs(form, rails.requiredInputSelector);
-        if (blankRequiredInputs && rails.fire(form, 'ajax:aborted:required', [blankRequiredInputs])) {
-          return rails.stopEverything(e);
-        }
-      }
+  if (rails.fire($document, 'rails:attachBindings')) {
 
-      if (remote) {
-        nonBlankFileInputs = rails.nonBlankInputs(form, rails.fileInputSelector);
-        if (nonBlankFileInputs) {
-          // slight timeout so that the submit button gets properly serialized
-          // (make it easy for event handler to serialize form without disabled values)
-          setTimeout(function(){ rails.disableFormElements(form); }, 13);
-          var aborted = rails.fire(form, 'ajax:aborted:file', [nonBlankFileInputs]);
+    $.ajaxPrefilter(function(options, originalOptions, xhr){ if ( !options.crossDomain ) { rails.CSRFProtection(xhr); }});
 
-          // re-enable form elements if event bindings return false (canceling normal form submission)
-          if (!aborted) { setTimeout(function(){ rails.enableFormElements(form); }, 13); }
+    // This event works the same as the load event, except that it fires every
+    // time the page is loaded.
+    //
+    // See https://github.com/rails/jquery-ujs/issues/357
+    // See https://developer.mozilla.org/en-US/docs/Using_Firefox_1.5_caching
+    $(window).on("pageshow.rails", function(){
+      $.rails.handlers.pageshow();
+    });
 
-          return aborted;
-        }
+    $document.delegate(rails.linkDisableSelector, 'ajax:complete', function() {
+      $.rails.handlers.ajaxCompleteLink($(this));
+    });
 
-        rails.handleRemote(form);
-        return false;
+    $document.delegate(rails.buttonDisableSelector, 'ajax:complete', function() {
+      $.rails.handlers.ajaxCompleteButton($(this));
```

Comment:
Maybe is better to only pass `this` instead of a jQuery object to keep the API consistent.